### PR TITLE
[Google Blockly] Re-implement unused blocks UI

### PR DESCRIPTION
**Pull request by ajpal**
This is the first customization we're re-implementing on Google's Blockly, so feedback on the general approach/structure is also encouraged! This PR also fixes the font on the blocks to be Gotham instead of the default sans-serif font.
Original [PR](https://github.com/code-dot-org/code-dot-org/pull/6875) and [spec](https://docs.google.com/document/d/1eclkN_U1lApiGDu9IgbPfnCmv0wvvkgBFZXgYSrBqvM/edit)

Before:
image

Google's default "disabled orphans" UI:
image

After:
image
image

And CDO Blockly for reference:
image

## Links

- [jira](https://codedotorg.atlassian.net/browse/STAR-1200)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/blockSvgUnused.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+  // We bind to mousedown rather than click so we can interrupt the drag
+  // that would otherwise be initiated.
+  Blockly.bindEvent_(this.frameHelp_, 'mousedown', this, function(e) {
+    if (Blockly.isRightButton(e)) {
+      // Right-click.
+      return;
+    }
+
+    this.helpClickFunc_(e);
```
Comment (ajpal) [inline apps/src/blockly/blockSvgUnused.js]: `Blockly.fireUiEvent` doesn't exist anymore, which I think is going to be kind of a bummer for a lot of our customizations. In this case, it was pretty straightforward to modify to be able to call the callback directly. Yet to be seen whether that works for other uses.

File: apps/src/blockly/blockSvgUnused.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
+import msg from '@cdo/locale';
```
Comment (ajpal) [inline apps/src/blockly/blockSvgUnused.js]: switched to pull from apps translation, not our custom Blockly strings

File: apps/src/blockly/blockSvgUnused.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+  this.frameHeader_ = Blockly.utils.dom.createSvgElement(
+    'rect',
+    {
+      x: -FRAME_MARGIN_SIDE,
+      y: -(FRAME_MARGIN_TOP + FRAME_HEADER_HEIGHT),
+      fill: '#c6cacd',
+      rx: 15,
+      ry: 15,
+      'clip-path': `url(#frameClip${this.block_.id.replace(/\)/g, '\\)')})`
```
Comment (ajpal) [inline apps/src/blockly/blockSvgUnused.js]: small fix to escape block IDs when necessary. Google Blockly block ids contain all of the following characters: `!#$%()*+,-./:;=?@[]^_``{|}~ ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789` [link](https://github.com/google/blockly/blob/93b5be40b6229cf164e83b63fe5a6cc4d44a9b5f/core/utils.js#L422)

I believe the only one that will cause issues here is `)`, but escaping it solves the issue

File: apps/src/blockly/blockSvgUnused.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    },
+    this.frameHelp_
+  );
+  Blockly.utils.dom
+    .createSvgElement(
+      'text',
+      {
+        class: 'blocklyText',
+        y: Blockly.ieVersion() ? 4 : 0, // again, offset text manually in IE
+        'dominant-baseline': 'central',
```
Comment (ajpal) [inline apps/src/blockly/blockSvgUnused.js]: add `dominant-baseline` and `text-anchor` styles. These appeared to be present on the element using CDO Blockly, but weren't set explicitly here, so my guess is they were applied using CSS at some other layer

File: apps/src/blockly/blockSvgUnused.js (hunk 2 above)
Comment (ajpal) [inline apps/src/blockly/blockSvgUnused.js]: Most of this file is copied wholesale from https://github.com/code-dot-org/blockly/blob/main/core/ui/block_svg/block_svg_unused.js, so I'll add some inline comments of relevant changes

File: apps/src/blockly/blockSvgUnused.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+      fill: '#c6cacd',
+      rx: 15,
+      ry: 15,
+      'clip-path': `url(#frameClip${this.block_.id.replace(/\)/g, '\\)')})`
+    },
+    this.frameGroup_
+  );
+
+  var frameTextVerticalPosition = -(FRAME_MARGIN_TOP + FRAME_HEADER_HEIGHT / 2);
+  if (Blockly.ieVersion()) {
```
Comment (breville) [inline apps/src/blockly/blockSvgUnused.js]: Is this still a reliable function?  Is there a form of "feature detection" that might be more reliable?

File: apps/src/blockly/blockSvgUnused.js (hunk 4 above)
Comment (breville) [reply apps/src/blockly/blockSvgUnused.js]: (Chrome inspector is useful for sourcing CSS on an element.)

Comment (ajpal) [reply apps/src/blockly/blockSvgUnused.js]: This comes from in CDO Blockly https://github.com/code-dot-org/blockly/blob/main/core/ui/css.js#L162

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 6]
```diff
@@ -146,6 +149,13 @@ function initializeBlocklyWrapper(blocklyInstance) {
   blocklyWrapper.wrapSettableProperty('valueTypeTabShapeMap');
   blocklyWrapper.wrapSettableProperty('Xml');
 
+  blocklyWrapper.ieVersion = () => false; // TODO
+
+  blocklyWrapper.isRightButton = function(e) {
+    // Control-clicking in WebKit on Mac OS X fails to change button to 2.
```
Comment (breville) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js]: macOS now :)

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 7]
```diff
@@ -210,9 +220,56 @@ function initializeBlocklyWrapper(blocklyInstance) {
...
+    if (this.unusedSvg_) {
+      this.unusedSvg_.dispose();
+      this.unusedSvg_ = null;
+    }
+  };
+  blocklyWrapper.BlockSvg.prototype.originalRender =
+    blocklyWrapper.BlockSvg.prototype.render;
+  blocklyWrapper.BlockSvg.prototype.render = function() {
+    this.originalRender();
+    this.removeUnusedFrame();
```
Comment (breville) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js]: Maybe some comments alluding to what an "unused frame" is?

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 8]
```diff
@@ -210,9 +220,56 @@ function initializeBlocklyWrapper(blocklyInstance) {
...
+
+  blocklyWrapper.blockRendering.PathObject.prototype.updateDisabled_ = function(
+    disabled
+  ) {
+    this.setClass_('blocklyDisabled', disabled);
+  };
+  blocklyWrapper.Workspace.prototype.addUnusedBlocksHelpListener = callback => {
+    Blockly.mainBlockSpace.addChangeListener(Blockly.Events.disableOrphans);
+    Blockly.bindEvent_(
+      Blockly.mainBlockSpace.getCanvas(),
```
Comment (breville) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js]: This does some interesting stuff, maybe a comment...

**Review by breville — APPROVED** [review]

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 9]
```diff
@@ -210,9 +213,70 @@ function initializeBlocklyWrapper(blocklyInstance) {
   blocklyWrapper.Workspace.prototype.getToolboxWidth = function() {
     return blocklyWrapper.mainBlockSpace.getMetrics().toolboxWidth;
   };
-  blocklyWrapper.Workspace.prototype.addUnusedBlocksHelpListener = () => {}; // TODO
-  blocklyWrapper.Workspace.prototype.getAllUsedBlocks =
-    blocklyWrapper.Workspace.prototype.getAllBlocks; // TODO
+
+  // Show Unused Code UI around unattached blocks
+  blocklyWrapper.BlockSvg.prototype.addUnusedBlockFrame = function(
```
Comment (Hamms) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js]: this pattern of modifying the prototypes of existing objects seems fragile at best. Do we have a long-term plan for how we want to handle extending google blockly functionality? Does google have recommendations/documentation somewhere for the expected way for third parties like ourselves to do stuff like this?

File: apps/src/sites/studio/pages/googleBlocklyWrapper.js [hunk 10]
```diff
@@ -210,9 +213,70 @@ function initializeBlocklyWrapper(blocklyInstance) {
...
+      this.unusedSvg_.dispose();
+      this.unusedSvg_ = null;
+    }
+  };
+  blocklyWrapper.BlockSvg.prototype.originalRender =
+    blocklyWrapper.BlockSvg.prototype.render;
+  blocklyWrapper.BlockSvg.prototype.render = function() {
+    this.originalRender();
+    this.removeUnusedBlockFrame();
+  };
```
Comment (Hamms) [inline apps/src/sites/studio/pages/googleBlocklyWrapper.js]: This pattern in particular seems really dangerous; it looks like we're attempting to implement inheritance without actually using it.

I'd be interested to see either a compositional implementation of this functionality, or one that actually uses property prototypal inheritance.

**Review by Hamms — CHANGES_REQUESTED** [review]
I've got some concerns with the way we're modifying blockly to make this change possible; in particular, I expect that the approach we're using here is going to make it significantly harder to make future changes.

**Comment by ajpal** [comment]
Updated screenshots:
image

image

File: apps/src/blocklyAddons/cdoTheme.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import GoogleBlockly from 'blockly/core';
+
+export default GoogleBlockly.Theme.defineTheme('modern', {
+  base: GoogleBlockly.Themes.Classic,
```
Comment (Hamms) [inline apps/src/blocklyAddons/cdoTheme.js:4]: nice!

**Comment by breville** [comment]
Great catch @Hamms!

**Review by Hamms — APPROVED** [review]
Love it!

As mentioned in DM, I do still have concerns about this being in our main project rather than in its own repo, but other than that I think this reimplementation looks good and I love the new inheritance model.

Would it be possible to get some tests for this and other CDO-specific functionality? Ideally here, but alternatively in a follow-up PR.

**Comment by ajpal** [comment]
> Would it be possible to get some tests for this and other CDO-specific functionality? Ideally here, but alternatively in a follow-up PR.

Good call on tests. So far, I don't have a way to run tests using Google Blockly. That is on my to-do list, but since there will be some setup required there, I will do it as follow up work.

**Pull request merged**